bsip-0064.md

```diff
@@ -0,0 +1,58 @@
+```
+BSIP: 64
+Title: Optional HTLC preimage length and HASH160 addition
+Authors: John Jones <jmjatlanta@gmail.com>, abitmore
+Status: Draft
+Type: Protocol
+Created: 2019-07-09
+Discussion: https://github.com/bitshares/bsips/issues/163
+```
+
+Abstract
+===================
+This BSIP proposes two changes to Hashed Time Lock Contracts (HTLC) defined in [BSIP 44](https://github.com/bitshares/bsips/blob/master/bsip-0044.md). The first makes the preimage `length` parameter optional. The second adds the HASH160 algorithm to the set of allowed hash functions.
+
+Motivation
+=================
+The original implementation of Hashed Timelock Contracts included a protection to prevent an attack based on the size of the preimage. While it is still believed this is a valuable feature, this makes the implementation incompatible with a strict implementation of the popular [BIP199](https://github.com/bitcoin/bips/blob/master/bip-0199.mediawiki) implemented by Bitcoin.
+
+In addition, BIP199 includes the ability to use HASH160, which is not currently part of the BitShares implementation.
+
+Risks
+===============
+**Note** below is a discussion of HTLC risks in general, not risks inherent in implementing this BSIP.
+
+**An Oversized Preimage Attack** There are limits to the size of a preimage that can be used. That limit is set by the blockchain itself. But when dealing with atomic swaps across blockchains, an attacker could use a preimage that is too large for one side of the 2 sided transaction. That means that an HTLC can be created on both chains, but only redeemed on one chain, and not the other.
+
+This can be mitigated by specifying the preimage size as part of the contract. This feature is available on many bitcoin-based chains among others, but is not standard. Should the preimage size be used with an atomic swap, both sides of the transaction should include the preimage size in their contract.
+
```

Comment:
Perhaps add a note that even if the preimage size is not part of the explicit contract, implicit limitations may still exist (e. g. maximum transaction size of the blockchain). Another risk is that an explicit limitation may be incompatible with the implicit one, without anyone noticing.